### PR TITLE
Release/1.10.34

### LINKED ISSUE #27: separated etcd cluster does not work out of the box
I have a script that shows this:

chef_etc_cluster_separate

will post more details later, when I've left it running for a couple of hours to prove.

### LINKED ISSUE #70: error: SDN node startup failed: clusternetwork "default" not found
Am running against origin/1.2.1, this config:

```
{
  "name": "ocp-cluster-environment",
  "description": "",
  "cookbook_versions": {},
  "json_class": "Chef::Environment",
  "chef_type": "environment",
  "default_attributes": {},
  "override_attributes": {
        "cookbook-openshift3": {
          "openshift_HA": true,
          "openshift_cluster_name": "master1.vagrant.test",
          "openshift_master_cluster_vip": "172.28.128.3",
          "openshift_deployment_type": "origin",
      "ose_version": "1.2.1-1.el7",
      "ose_major_version": "1.2",
"openshift_common_use_openshift_sdn": false,
[...node data...]
```

Does this error look familiar to anyone?

```
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.393260   67368 plugins.go:71] No cloud provider specified.
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.393348   67368 common.go:77] Initializing single-tenant plugin for master1.vagrant.test (172.28.128.3)
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.393371   67368 start_node.go:288] Starting node master1.vagrant.test (v1.2.1)
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.398676   67368 start_node.go:297] Connecting to API server https://master1.vagrant.test:8443
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.401893   67368 node.go:131] Connecting to Docker at unix:///var/run/docker.sock
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: I0207 15:57:57.476270   67368 manager.go:132] cAdvisor running in container: "/"
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: E0207 15:57:57.478323   67368 common.go:192] Failed to obtain ClusterNetwork: clusternetwork "default" not found
Feb 07 15:57:57 master1.vagrant.test origin-node[67368]: F0207 15:57:57.478404   67368 node.go:310] error: SDN node startup failed: clusternetwork "default" not found
```

I think I remember seeing it before but can't recall the solution, googling came up with nothing.

### PR DESCRIPTION
**Pull request by IshentRas**
Fix #27 
Fix #70

### CODE REVIEW

**Comment by IshentRas** [comment]
Thanks @jperville I do believe we are getting better :+1:

File: providers/openshift_add_etcd.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+def whyrun_supported?
+  true
+end
+
+action :add_node do
+  etcd_servers = new_resource.etcd_servers
+  etcd_servers.each do |etcd|
+    bash 'add_etcd_nodes_if_needed' do
+      code <<-EOH
+        liste_nodes=$(etcdctl -C https://${ETCD_SERVER}:${ETCD_CLIENT_PORT} --cert-file $CRT --key-file $KEY --ca-file $CA member list | cut -d' ' -f2)
```
Comment (ianmiell) [inline providers/openshift_add_etcd.rb]: liste - do i need to learn french now :) ?

File: recipes/etcd_cluster.rb [hunk 2]
```diff
@@ -145,7 +150,12 @@
   template "#{node['cookbook-openshift3']['etcd_conf_dir']}/etcd.conf" do
     source 'etcd.conf.erb'
     notifies :restart, 'service[etcd]', :immediately
-    variables etcd_servers: etcd_servers
+    variables lazy {
```
Comment (ianmiell) [inline recipes/etcd_cluster.rb]: this is a new technique for me - do these only get assigned if lazy is referenced?

File: README.md [hunk 3]
```diff
@@ -369,6 +369,32 @@ In general, override attributes in the environment should be used when changing
       ]
 ...
 ```
+* REMOVE ETCD SERVERS FROM CLUSTER ("etcd_removal_servers" must be define and list all servers you want to remove. etcd_servers should be your desire state"
+
+```json
+...
+      "etcd_removal_servers": [
```
Comment (ianmiell) [inline README.md]: remove_servers would be better.

Comment (ianmiell) [inline README.md]: ie etcd_remove_servers

File: providers/openshift_add_etcd.rb [hunk 4]
```diff
@@ -36,3 +36,30 @@ def whyrun_supported?
...
   new_resource.updated_by_last_action(true)
 end
+
+action :remove_node do
+  etcd_servers = new_resource.etcd_servers
+  etcd_servers_to_remove = new_resource.etcd_servers_to_remove
+  etcd_servers_to_remove.each do |etcd|
+    bash 'remove_etcd_nodes_if_needed' do
+      code <<-EOH
+        liste_nodes=$(etcdctl -C https://${ETCD_SERVER}:${ETCD_CLIENT_PORT} --cert-file $CRT --key-file $KEY --ca-file $CA member list | cut -d' ' -f2)
```
Comment (ianmiell) [inline providers/openshift_add_etcd.rb]: liste

File: recipes/etcd_cluster.rb [hunk 5]
```diff
@@ -165,6 +165,13 @@
     }
   end
 
+  directory "#{node['cookbook-openshift3']['etcd_data_dir']}member" do
```
Comment (ianmiell) [inline recipes/etcd_cluster.rb]: /member ?

File: recipes/etcd_cluster.rb (hunk 2 above)
Comment (IshentRas) [inline recipes/etcd_cluster.rb]: https://docs.chef.io/resource_common.html#lazy-evaluation

File: README.md (hunk 3 above)
Comment (IshentRas) [inline README.md]: Makes sense :+1:

File: providers/openshift_add_etcd.rb (hunk 4 above)
Comment (IshentRas) [inline providers/openshift_add_etcd.rb]: French issue :-)

File: recipes/etcd_cluster.rb (hunk 5 above)
Comment (IshentRas) [inline recipes/etcd_cluster.rb]: I put that logic to clear a pre-existing etcd member. But you are right it should not be there...

Comment (IshentRas) [inline recipes/etcd_cluster.rb]: Based on the your comments #27

File: providers/openshift_add_etcd.rb (hunk 1 above)
Comment (IshentRas) [inline providers/openshift_add_etcd.rb]: Fixed

File: providers/openshift_add_etcd.rb (hunk 4 above)
Comment (IshentRas) [inline providers/openshift_add_etcd.rb]: Fixed

**Comment by IshentRas** [comment]
Please review and merge

**Pull request merged**